compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1468,6 +1545,15 @@ public void visit(BLangAssignment assignNode) {
         setTypeOfVarRefInAssignment(varRef);
         expType = varRef.type;
 
+        if (varRef.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+            BLangSimpleVarRef variableRef = (BLangSimpleVarRef) varRef;
+            if (variableRef.lhsVar && variableRef.symbol != null &&
+                    (variableRef.symbol.tag & SymTag.TYPE_DEF) == SymTag.TYPE_DEF) {
```

Comment:
Need to do this for all module level final constructs like functions too. We can fix that separately maybe. Let's create an issue.

Reply:
Created https://github.com/ballerina-platform/ballerina-lang/issues/30577